File: src/medium-98/iteration.ts
import { BinaryTreeNode } from 'data-structures/binary-tree'

export function isValidBST(node: BinaryTreeNode<number> | null) {
  const stack: (BinaryTreeNode<number> | null)[] = []
  let current: BinaryTreeNode<number> | null = node
  let prev: number = -Infinity

  while (current !== null || stack.length > 0) {
    while (current !== null) {
      stack.push(current)
      current = current.left
    }
    current = stack.pop()!
    if (current.value <= prev) return false
    prev = current.value
    current = current.right
  }

  return true
}

File: src/medium-98/iteration.go
package medium98

import (
	structures "leetcode-solutions/data-structures"
	"math"
)

type node = structures.BinaryTreeNode

func isValidBSTByIteration(root *node) bool {
	if root == nil {
		return true
	}

	stack, current, prevNodeVal := []*node{}, root, math.MinInt32
	for current != nil || len(stack) > 0 {
		for current != nil {
			stack = append(stack, current)
			current = current.Left
		}
		topFrameIndex := len(stack) - 1
		current = stack[topFrameIndex]
		stack = stack[:topFrameIndex]

		if current.Val <= prevNodeVal {
			return false
		}
		prevNodeVal = current.Val

		current = current.Right
	}
	return true
}
